server/src-lib/Hasura/RQL/DML/Select/Internal.hs

```diff
@@ -153,6 +153,25 @@ buildJsonObject pfx parAls arrRelCtx flds =
                      ANIField (fldAls, arrSel)
         in S.mkQIdenExp arrPfx fldAls
 
+    toSQLCol :: PGColInfo -> S.SQLExp
+    toSQLCol col = toJSONableExp (pgiType col) $
+                    S.mkQIdenExp (mkBaseTableAls pfx) $ pgiName col
+
+    toSQLColArgs :: PGColInfo -> AnnArgs -> S.SQLExp
+    toSQLColArgs col args
+      | isJSONType colTy || isJSONBType colTy =
+          case (getJSONArgumentPath args) of
+            Nothing -> toSQLCol col
+            Just pArg -> toJSONableExp colTy $
+                          S.SEOpApp S.jsonbGetOp
+                            [ S.mkQIdenExp (mkBaseTableAls pfx) $ pgiName col
```

Comment:
The logic here would be pretty straight forward if we use `ColOp` kind of data type.